[904] Enable testing component JS interactions using new with_rendered_component_path method (#1061)
ViewComponent/view_component: docs/guide/testing.md

# config/application.rb
config.view_component.preview_paths << "#{Rails.root}/spec/components/previews"
```

## Testing Interactive Components

Use the `visit_rendered_component_in_browser` helper method in your system tests to test interactivity of components. 

```rb
require "test_helper"

class ViewComponentSystemTest < ViewComponent::SystemTestCase
  driven_by :cuprite

  def test_simple_js_interaction_in_browser_without_layout
    visit_rendered_component_in_browser(SimpleJavascriptInteractionWithJsIncludedComponent.new)

    assert find("[data-hidden-field]", visible: false)
    find("[data-button]", text: "Click Me To Reveal Something Cool").click
    assert find("[data-hidden-field]", visible: true)
  end
end
```

### Handling components without JS included

> A common pattern is to include assets in a layout rather than directly inside the component. In those cases, you can define a `layout` in the `visit_rendered_component_in_browser` so that the component is rendered with those assets:

Reviewer: Would it be possible to add a configuration option for the default layout to use?

# config/application.rb
config.view_component.preview_paths << "#{Rails.root}/spec/components/previews"
```

> ## Testing Interactive Components

Author: @joelhawksley added this documentation. Let me know if this is unclear. Thank you!

# config/application.rb
config.view_component.preview_paths << "#{Rails.root}/spec/components/previews"
```

## Testing Interactive Components

> Use the `visit_rendered_component_in_browser` helper method in your system tests to test interactivity of components.

Reviewer 2: Just a note to update the docs in general ❤️

# config/application.rb
config.view_component.preview_paths << "#{Rails.root}/spec/components/previews"
```

> ## Testing Interactive Components

Reviewer 2: ```suggestion
## Component system tests
```

> Use `with_rendered_component_path` with `render_inline` in system tests to test interactive components:

Reviewer 2: ```suggestion
Use `with_rendered_component_path` with `render_inline` to system test components:
```

```rb
class ViewComponentSystemTest < ViewComponent::SystemTestCase
  def test_simple_js_interaction_in_browser_without_layout
    with_rendered_component_path(render_inline(SimpleJavascriptInteractionWithJsIncludedComponent.new)) do |path|
      visit(path)

      assert(find("[data-hidden-field]", visible: false))
      find("[data-button]", text: "Click Me To Reveal Something Cool").click
      assert(find("[data-hidden-field]", visible: true))
    end
  end
end
```

> For components that depend on JavaScript provided by a layout, provide the `layout` argument:

Reviewer 2: ```suggestion
For components that depend on a layout, provide the `layout` argument:
```

```rb
class ViewComponentSystemTest < ViewComponent::SystemTestCase
  def test_simple_js_interaction_in_browser_with_layout
    with_rendered_component_path(render_inline(SimpleJavascriptInteractionWithoutJsIncludedComponent.new), layout: 'application') do |path|
      visit(path)

>       assert(find("[data-hidden-field]", visible: false))
>       find("[data-button]", text: "Click Me To Reveal Something Cool").click
>       assert(find("[data-hidden-field]", visible: true))

Reviewer 2: ```suggestion
      # ...
```

# config/application.rb
config.view_component.preview_paths << "#{Rails.root}/spec/components/previews"
```

## Testing Interactive Components

Use `with_rendered_component_path` in system tests to test interactive components:

```rb
class ViewComponentSystemTest < ViewComponent::SystemTestCase

>   driven_by :cuprite
> 

Reviewer 2: ```suggestion
```

  def test_simple_js_interaction_in_browser_without_layout
    with_rendered_component_path(SimpleJavascriptInteractionWithJsIncludedComponent.new) do |path|
      visit(path)

      assert(find("[data-hidden-field]", visible: false))
      find("[data-button]", text: "Click Me To Reveal Something Cool").click
      assert(find("[data-hidden-field]", visible: true))
    end
  end
end
```

### Testing components with external JavaScript

For components that depend on JavaScript provided by a layout, provide the `layout` argument:

```rb
class ViewComponentSystemTest < ViewComponent::SystemTestCase

>   driven_by :cuprite
> 

Reviewer 2: ```suggestion
```

# config/application.rb
config.view_component.preview_paths << "#{Rails.root}/spec/components/previews"
```

## Testing Interactive Components

> Use the `with_rendered_component_path` helper method in your system tests to test interactivity of components.

Reviewer 2: ```suggestion
Use `with_rendered_component_path` in system tests to test interactive components:
```

> Use the `with_rendered_component_in_browser` helper method in your system tests to test interactivity of components.

Reviewer 2: ```suggestion
Use the `with_rendered_component_path` helper method in system tests:
```

```rb
require "test_helper"

class ViewComponentSystemTest < ViewComponent::SystemTestCase
  driven_by :cuprite

  def test_simple_js_interaction_in_browser_without_layout

>     with_rendered_component_in_browser(SimpleJavascriptInteractionWithJsIncludedComponent.new) do |page|

Reviewer 2: So this would be the API:

```suggestion
    with_rendered_component_path(SimpleJavascriptInteractionWithJsIncludedComponent.new) do |path|
```

>     with_rendered_component_path(SimpleJavascriptInteractionWithJsIncludedComponent.new) do |page|
>       visit page

Reviewer 2: `page` is a reserved word in most system tests:

```suggestion
    with_rendered_component_path(SimpleJavascriptInteractionWithJsIncludedComponent.new) do |path|
      visit(path)
```

Author: Good point!

>       assert find("[data-hidden-field]", visible: false)
>       find("[data-button]", text: "Click Me To Reveal Something Cool").click
>       assert find("[data-hidden-field]", visible: true)

Reviewer 2: ```suggestion
      assert(find("[data-hidden-field]", visible: false))
      find("[data-button]", text: "Click Me To Reveal Something Cool").click
      assert(find("[data-hidden-field]", visible: true))
```

    end
  end
end
```

> ### Handling components without JS included

Reviewer 2: ```suggestion
### Testing components with external JavaScript
```

> A common pattern is to include assets in a layout rather than directly inside the component. In those cases, you can define a `layout` in the `visit_rendered_component_in_browser` so that the component is rendered with those assets:

Reviewer 2: ```suggestion
For components that depend on JavaScript provided by a layout, provide the `layout` argument:
```

```rb
require "test_helper"

class ViewComponentSystemTest < ViewComponent::SystemTestCase
  driven_by :cuprite

  def test_simple_js_interaction_in_browser_with_layout

>     with_rendered_component_path(SimpleJavascriptInteractionWithoutJsIncludedComponent.new, layout: 'application') do |page|
>       visit page

Reviewer 2: ```suggestion
    with_rendered_component_path(SimpleJavascriptInteractionWithoutJsIncludedComponent.new, layout: 'application') do |path|
      visit(path)
```

>       assert find("[data-hidden-field]", visible: false)
>       find("[data-button]", text: "Click Me To Reveal Something Cool").click
>       assert find("[data-hidden-field]", visible: true)

Reviewer 2: ```suggestion
      assert(find("[data-hidden-field]", visible: false))
      find("[data-button]", text: "Click Me To Reveal Something Cool").click
      assert(find("[data-hidden-field]", visible: true))
```
